### PR TITLE
LLVM 16 + Features

### PR DESCRIPTION
**Pull request by m-kurtenacker**
Moves to llvm 16.
Additionally merges:
* JSON backend to work with anyopt.
* Newmem, adding the memory monad to branching terminators.
* unify_def_type.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -3,7 +3,7 @@ cmake_minimum_required(VERSION 3.13.4 FATAL_ERROR)
 project(Thorin)
 
 set(PACKAGE_VERSION "0.3.9")
-set(CMAKE_CONFIGURATION_TYPES "Debug;Release" CACHE STRING "limited config" FORCE)
+#set(CMAKE_CONFIGURATION_TYPES "Debug;Release" CACHE STRING "limited config" FORCE)
```
Comment (Hugobros3) [inline CMakeLists.txt:6]: What's that about ?

File: src/thorin/analyses/schedule.cpp [hunk 2]
```diff
@@ -9,6 +9,8 @@
 #include "thorin/analyses/looptree.h"
 #include "thorin/analyses/scope.h"
 
+#include <stack>
```
Comment (Hugobros3) [inline src/thorin/analyses/schedule.cpp:12]: didn't we revert 2c0ce6c68db2ebd70ec40c4c3229cd04f2ea7ad1 on development ?

File: src/thorin/be/c/c.cpp [hunk 3]
```diff
@@ -358,7 +358,7 @@ void CCodeGen::emit_module() {
     Scope::for_each(world(), [&] (const Scope& scope) {
         if (scope.entry()->name() == "hls_top")
             hls_top = scope.entry();
-        else
+        else if (scope.entry()->cc() != CC::Internal)
```
Comment (Hugobros3) [inline src/thorin/be/c/c.cpp:361]: why ? doesn't this just break the codegen by leaving fns out ?

File: src/thorin/continuation.h [hunk 4]
```diff
@@ -72,16 +76,18 @@ class App : public Def {
...
 
 //------------------------------------------------------------------------------
 
 enum class CC : uint8_t {
-    C,          ///< C calling convention.
-    Device,     ///< Device calling convention. These are special functions only available on a particular device.
+    C,              ///< C calling convention.
+    Device,         ///< Device calling convention. These are special functions only available on a particular device.
+    DeviceHostCode, ///< Calling convention to denote continuations that are generated as device code.
+    Internal,       ///< External, but only for linking with artic or anyopt.
```
Comment (Hugobros3) [inline src/thorin/continuation.h:90]: why is there an "internal" CC ? if the continuation is internal it effectively doesn't matter, or does it ?

File: src/thorin/debug.h [hunk 5]
```diff
@@ -43,18 +43,35 @@ class Debug {
...
+        , creation_context("")
+#endif
         , loc(loc)
         , meta(meta)
     {}
     Debug(const char* name, Loc loc = {}, const Def* meta = nullptr)
         : Debug(std::string(name), loc, meta)
     {}
+#if THORIN_ENABLE_CREATION_CONTEXT
+    Debug(std::string name, std::string creation_context, Loc loc = {}, const Def* meta = nullptr)
```
Comment (Hugobros3) [inline src/thorin/debug.h:56]: what's the idea here?

File: src/thorin/util/hash.h [hunk 6]
```diff
@@ -217,6 +217,7 @@ class HashTable {
     HashTable()
         : capacity_(StackCapacity)
         , size_(0)
+        , array_()
```
Comment (Hugobros3) [inline src/thorin/util/hash.h:220]: ?

File: src/thorin/be/emitter.h [hunk 7]
```diff
@@ -1,6 +1,9 @@
 #ifndef THORIN_BE_EMITTER_H
 #define THORIN_BE_EMITTER_H
 
+#include<stack>
```
Comment (Hugobros3) [inline src/thorin/be/emitter.h:4]: Same concern as before

File: src/thorin/be/json/json.cpp [hunk 8]
```diff
@@ -0,0 +1,671 @@
+#include "json.h"
+
+namespace thorin::json {
+
+class TypeTable {
```
Comment (Hugobros3) [inline src/thorin/be/json/json.cpp:5]: Since we merged Def and Type, should this be refactored ?

File: CMakeLists.txt (hunk 1 above)
Comment (PearCoding) [reply CMakeLists.txt:6]: Never understood why it was there the first place. Usually the default are "Debug;MinSize;RelWithDebug;Release". But we limited ourselfs to Debug and Release. So happy to see it go finally. :D

**Pull request closed without merging**